File: sdks/kotlin/src/test/kotlin/dev/wheso/TraceShardTest.kt
// トレースベクタの照合（Kotlin、層 2: 決定同一）。
//
// 凍結トレース（spec/vectors/trace-shard.jsonl）を Kotlin の判断コアへ流し、
// 出力コマンド列が TypeScript の参照実装と**完全に一致**することを確かめる。
// 1 コマンドの相違も許さない（conformance.md 4.4）。
//
// 相違した場合はベクタではなく実装を直す（ADR-0012）。
package dev.wheso

import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

class TraceShardTest {
    private fun longOf(obj: JsonObject, key: String): Long {
        val value = obj[key]?.jsonPrimitive?.content
        return value?.toLongOrNull() ?: error("$key が整数ではない")
    }

    private fun toEvent(input: JsonObject): ShardEvent {
        return when (input["kind"]?.jsonPrimitive?.content) {
            "media" -> ShardEvent.Media(
                from = longOf(input, "from"),
                ch = longOf(input, "ch"),
                sid = longOf(input, "sid"),
                tid = longOf(input, "tid"),
                key = input["key"]?.jsonPrimitive?.content == "true",
                bytes = longOf(input, "bytes"),
                flags = longOf(input, "flags"),
            )
            "subscribe" -> ShardEvent.Subscribe(
                from = longOf(input, "from"),
                to = longOf(input, "to"),
                want = input["want"]?.jsonPrimitive?.content == "true",
                maxSpatialId = longOf(input, "maxSpatialId"),
            )
            "join" -> ShardEvent.Join(longOf(input, "id"))
            "leave" -> ShardEvent.Leave(longOf(input, "id"))
            "link" -> ShardEvent.Link(
                peer = input["peer"]?.jsonPrimitive?.content?.toLongOrNull() ?: 0L,
                state = input["state"]?.jsonPrimitive?.content ?: "",
            )
            "timer" -> ShardEvent.Timer
            "budget" -> ShardEvent.Budget(longOf(input, "bytesPerSec"))
            "report" -> {
                val samples = input["delayUs"]?.jsonArray ?: error("delayUs が無い")
                ShardEvent.Report(
                    from = longOf(input, "from"),
                    delayUs = samples.map { it.jsonPrimitive.content.toLongOrNull() ?: error("整数でない") },
                )
            }
            else -> error("未知のイベント: ${input["kind"]}")
        }
    }

    /** 出力コマンドを TypeScript と同じ JSON 表現へ写す。欄名も一致させる。 */
    private fun toJson(command: ShardCommand): JsonObject = when (command) {
        is ShardCommand.Forward -> buildJsonObject {
            put("kind", JsonPrimitive("forward"))
            put("to", buildJsonArray { command.to.forEach { add(JsonPrimitive(it)) } })
        }
        is ShardCommand.Drop -> buildJsonObject {
            put("kind", JsonPrimitive("drop"))
            put("priority", JsonPrimitive(command.priority))
            put("count", JsonPrimitive(command.count))
        }
        is ShardCommand.Notify -> buildJsonObject {
            put("kind", JsonPrimitive("notify"))
            put("code", JsonPrimitive(command.code))
        }
        is ShardCommand.SetTier -> buildJsonObject {
            put("kind", JsonPrimitive("setTier"))
            put("for", JsonPrimitive(command.targetId))
            put("tier", JsonPrimitive(command.tier))
        }
    }

    /** 期待値の並びを、欄の順序に依存しない形で比べる。 */
    private fun normalize(array: JsonArray): List<Map<String, String>> =
        array.map { element ->
            element.jsonObject.entries.associate { (key, value) ->
                key to (if (value is JsonArray) value.toString() else value.jsonPrimitive.content)
            }
        }

    @Test
    fun frozenShardTraceMatchesTypescriptReference() {
        val lines = File("../../spec/vectors/trace-shard.jsonl")
            .readLines()
            .filter { it.isNotBlank() }
        assertTrue(lines.size > 100, "トレースが十分な行数を持つ")

        val header = Json.parseToJsonElement(lines.first()).jsonObject
        assertEquals("shard", header["unit"]?.jsonPrimitive?.content, "中継ノードのトレースである")

        var state: ShardState? = null
        var pending: JsonObject? = null
        var checked = 0

        for (line in lines.drop(1)) {
            val row = Json.parseToJsonElement(line).jsonObject
            val input = row["in"]
            if (input != null) {
                pending = input.jsonObject
                continue
            }
            val out = row["out"] ?: continue
            val event = pending ?: error("出力に対応する入力が無い")
            pending = null
            val t = longOf(row, "t")
            // 初期状態の時刻はトレースの最初の t と一致させる必要がある。
            val current = state ?: initialShardState(t)
            val result = shardStep(current, toEvent(event), t)
            state = result.state
            val actual = buildJsonArray { result.commands.forEach { add(toJson(it)) } }
            assertEquals(
                normalize(out.jsonArray),
                normalize(actual),
                "入力 $event に対する出力が一致する",
            )
            checked += 1
        }
        assertTrue(checked > 100, "十分な行数を照合した（実際 $checked）")
    }
}

File: sdks/dart/test/trace_shard_test.dart
// トレースベクタの照合（Dart、層 2: 決定同一）。
//
// 凍結トレース（spec/vectors/trace-shard.jsonl）を Dart の判断コアへ流し、
// 出力コマンド列が TypeScript の参照実装と**完全に一致**することを確かめる。
// 1 コマンドの相違も許さない（conformance.md 4.4）。
//
// 相違した場合はベクタではなく実装を直す（ADR-0012）。
//
// 動的型（dynamic）を使わない。JSON は Object? で受け、is で絞る。

import 'dart:convert';
import 'dart:io';

import 'package:test/test.dart';
import 'package:wheso_client/src/shard_core.dart';

/// 連想配列として読む。型が違えば試験を失敗させる。
Map<String, Object?> readMap(Object? value, String where) {
  if (value is Map<String, Object?>) {
    return value;
  }
  throw StateError('$where: 連想配列ではない');
}

List<Object?> readList(Object? value, String where) {
  if (value is List<Object?>) {
    return value;
  }
  throw StateError('$where: 配列ではない');
}

int readInt(Map<String, Object?> map, String key, String where) {
  final value = map[key];
  if (value is int) {
    return value;
  }
  throw StateError('$where: $key が整数ではない');
}

bool readBool(Map<String, Object?> map, String key, String where) {
  final value = map[key];
  if (value is bool) {
    return value;
  }
  throw StateError('$where: $key が真偽値ではない');
}

/// 入力行を ShardEvent に変換する。
ShardEvent toEvent(Map<String, Object?> input) {
  final kind = input['kind'];
  switch (kind) {
    case 'media':
      return MediaEvent(
        from: readInt(input, 'from', 'media'),
        ch: readInt(input, 'ch', 'media'),
        sid: readInt(input, 'sid', 'media'),
        tid: readInt(input, 'tid', 'media'),
        key: input['key'] == true,
        bytes: readInt(input, 'bytes', 'media'),
        flags: readInt(input, 'flags', 'media'),
        seq: readInt(input, 'seq', 'media'),
      );
    case 'subscribe':
      return SubscribeEvent(
        from: readInt(input, 'from', 'subscribe'),
        to: readInt(input, 'to', 'subscribe'),
        ch: readInt(input, 'ch', 'subscribe'),
        want: readBool(input, 'want', 'subscribe'),
        maxSpatialId: readInt(input, 'maxSpatialId', 'subscribe'),
        maxTemporalId: readInt(input, 'maxTemporalId', 'subscribe'),
      );
    case 'ack':
      return AckEvent(
        from: readInt(input, 'from', 'ack'),
        to: readInt(input, 'to', 'ack'),
        ch: readInt(input, 'ch', 'ack'),
        sid: readInt(input, 'sid', 'ack'),
        highestSeq: readInt(input, 'highestSeq', 'ack'),
      );
    case 'streamAnnounce':
      final rungsList = readList(input['rungs'], 'streamAnnounce rungs');
      final List<LadderRung> rungs = [];
      for (final item in rungsList) {
        final r = readMap(item, 'streamAnnounce rung');
        rungs.add(LadderRung(
          sid: readInt(r, 'sid', 'rung'),
          width: readInt(r, 'width', 'rung'),
          height: readInt(r, 'height', 'rung'),
          framerate: readInt(r, 'framerate', 'rung'),
          temporalLayers: readInt(r, 'temporalLayers', 'rung'),
          targetBitrate: readInt(r, 'targetBitrate', 'rung'),
        ));
      }
      return StreamAnnounceEvent(
        from: readInt(input, 'from', 'streamAnnounce'),
        ch: readInt(input, 'ch', 'streamAnnounce'),
        rungs: rungs,
      );
    case 'join':
      return JoinEvent(id: readInt(input, 'id', 'join'));
    case 'leave':
      return LeaveEvent(id: readInt(input, 'id', 'leave'));
    case 'link':
      final peer = input['peer'];
      final state = input['state'];
      return LinkEvent(
        peer: peer is int ? peer : 0,
        state: state is String ? state : '',
      );
    case 'timer':
      return const TimerEvent();
    case 'budget':
      return BudgetEvent(bytesPerSec: readInt(input, 'bytesPerSec', 'budget'));
    case 'report':
      final samples = readList(input['delayUs'], 'report delayUs');
      final List<int> delayUs = [];
      for (final sample in samples) {
        if (sample is! int) {
          throw StateError('report delayUs: 整数ではない');
        }
        delayUs.add(sample);
      }
      return ReportEvent(from: readInt(input, 'from', 'report'), delayUs: delayUs);
    case 'keyframeRequest':
      return KeyframeRequestEvent(
        from: readInt(input, 'from', 'keyframeRequest'),
        target: readInt(input, 'target', 'keyframeRequest'),
        ch: readInt(input, 'ch', 'keyframeRequest'),
        sid: readInt(input, 'sid', 'keyframeRequest'),
      );
  }
  throw StateError('未知のイベント: $kind');
}

/// 出力コマンドを TypeScript と同じ JSON 表現へ写す。欄名も順序も一致させる。
Map<String, Object?> toJson(ShardCommand command) {
  switch (command) {
    case ForwardCommand():
      return <String, Object?>{'kind': 'forward', 'to': command.to};
    case DropCommand():
      return <String, Object?>{'kind': 'drop', 'priority': command.priority, 'count': command.count};
    case SetTierCommand():
      return <String, Object?>{'kind': 'setTier', 'for': command.targetId, 'tier': command.tier};
    case KeyframeRequestCommand():
      return <String, Object?>{
        'kind': 'keyframeRequest',
        'for': command.targetId,
        'channel': command.channel,
        'spatialId': command.spatialId,
      };
    case AckUpstreamCommand():
      return <String, Object?>{
        'kind': 'ackUpstream',
        'to': command.to,
        'channel': command.channel,
        'spatialId': command.spatialId,
        'highestSeq': command.highestSeq,
      };
    case DisconnectCommand():
      return <String, Object?>{'kind': 'disconnect', 'peer': command.peer};
    case NotifyCommand():
      return <String, Object?>{'kind': 'notify', 'code': command.code};
  }
}

void main() {
  test('中継ノードの凍結トレースが参照実装と完全一致する', () {
    final file = File('../../spec/vectors/trace-shard.jsonl');
    final lines = file.readAsLinesSync().where((line) => line.trim().isNotEmpty).toList();
    expect(lines.length, greaterThan(100));

    final header = readMap(jsonDecode(lines.first), 'header');
    expect(header['unit'], equals('shard'));

    ShardState? state;
    Map<String, Object?>? pending;
    var checked = 0;

    for (final line in lines.skip(1)) {
      final row = readMap(jsonDecode(line), 'row');
      final input = row['in'];
      if (input != null) {
        pending = readMap(input, 'in');
        continue;
      }
      final out = row['out'];
      if (out == null) {
        continue;
      }
      final expected = readList(out, 'out');
      final event = pending;
      if (event == null) {
        throw StateError('出力に対応する入力が無い');
      }
      pending = null;
      final t = readInt(row, 't', 'row');
      state ??= initialState(t);
      final result = step(state, toEvent(event), t);
      state = result.state;
      final actual = result.commands.map(toJson).toList();
      expect(jsonEncode(actual), equals(jsonEncode(expected)),
          reason: '入力 ${jsonEncode(event)} に対する出力が一致する');
      checked += 1;
    }
    expect(checked, greaterThan(100), reason: '十分な行数を照合した');
  });
}
